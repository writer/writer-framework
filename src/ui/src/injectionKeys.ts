import { ComputedRef, InjectionKey, Ref, VNode } from "vue";
import {
	BuilderManager,
	CollaborationManager,
	Component,
	Core,
	InstancePath,
	InstancePathItem,
	NotesManager,
	SecretsManager,
} from "./writerTypes";

export default {
	core: Symbol() as InjectionKey<Core>,
	builderManager: Symbol() as InjectionKey<BuilderManager>,
	notesManager: Symbol() as InjectionKey<NotesManager>,
	collaborationManager: Symbol() as InjectionKey<CollaborationManager>,
	secretsManager: Symbol() as InjectionKey<SecretsManager>,
	evaluatedFields: Symbol() as InjectionKey<Record<string, ComputedRef<any>>>,
	componentId: Symbol() as InjectionKey<Component["id"]>,
	isBeingEdited: Symbol() as InjectionKey<Ref<boolean>>,
	isDisabled: Symbol() as InjectionKey<Ref<boolean>>,
	getChildrenVNodes: Symbol() as InjectionKey<
		(instanceNumber?: InstancePathItem["instanceNumber"]) => VNode[]
	>,
	renderProxiedComponent: Symbol() as InjectionKey<
		(
			componentId: Component["id"],
			instanceNumber: InstancePathItem["instanceNumber"],
			ext?: { class?: string[]; contextSlot?: string },
		) => VNode
	>,
	instancePath: Symbol() as InjectionKey<InstancePath>,
	flattenedInstancePath: Symbol() as InjectionKey<string>,
	instanceData: Symbol() as InjectionKey<Ref[]>,
};
