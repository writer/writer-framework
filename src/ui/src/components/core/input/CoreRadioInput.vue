<template>
	<BaseInputWrapper
		ref="rootInstance"
		:label="fields.label.value"
		class="CoreRadioInput"
	>
		<div
			class="options"
			:class="{
				horizontal: fields.orientation.value == 'horizontal',
			}"
		>
			<div
				v-for="(option, optionKey) in fields.options.value"
				:key="optionKey"
				class="option"
			>
				<input
					:id="`${flattenedInstancePath}-option-${optionKey}`"
					type="radio"
					:checked="formValue === optionKey"
					:value="optionKey"
					:name="`${flattenedInstancePath}-options`"
					@input="
						($event) => handleInput(optionKey, 'wf-option-change')
					"
				/><label
					:for="`${flattenedInstancePath}-option-${optionKey}`"
					>{{ option }}</label
				>
			</div>
		</div>
	</BaseInputWrapper>
</template>

<script lang="ts">
import { inject } from "vue";
import { ref } from "vue";
import { FieldCategory, FieldType } from "@/writerTypes";
import {
	accentColor,
	cssClasses,
	primaryTextColor,
} from "@/renderer/sharedStyleFields";
import BaseInputWrapper from "../base/BaseInputWrapper.vue";
import { ComponentPublicInstance } from "vue";
import { validatorObjectRecordNotNested } from "@/constants/validators";

const description =
	"A user input component that allows users to choose a single value from a list of options using radio buttons.";

const defaultOptions = { a: "Option A", b: "Option B" };

const onChangeHandlerStub = `
def onchange_handler(state, payload):

	# Set the state variable "selected" to the selected radio option

	state["selected"] = payload`;

export default {
	writer: {
		name: "Radio Input",
		description,
		category: "Input",
		fields: {
			label: {
				name: "Label",
				init: "Input Label",
				type: FieldType.Text,
			},
			options: {
				name: "Options",
				desc: "Key-value object with options. Must be a JSON string or a state reference to a dictionary.",
				type: FieldType.KeyValue,
				default: JSON.stringify(defaultOptions, null, 2),
				validator: validatorObjectRecordNotNested,
			},
			orientation: {
				name: "Orientation",
				type: FieldType.Text,
				options: {
					vertical: "Vertical",
					horizontal: "Horizontal",
				},
				default: "vertical",
				category: FieldCategory.Style,
				desc: "Specify how to lay out the options.",
			},
			primaryTextColor,
			accentColor,
			cssClasses,
		},
		events: {
			"wf-option-change": {
				desc: "Sent when the selected option changes.",
				stub: onChangeHandlerStub.trim(),
				bindable: true,
			},
		},
	},
};
</script>

<script setup lang="ts">
import injectionKeys from "@/injectionKeys";
import { useFormValueBroker } from "@/renderer/useFormValueBroker";

const fields = inject(injectionKeys.evaluatedFields);
const flattenedInstancePath = inject(injectionKeys.flattenedInstancePath);
const rootInstance = ref<ComponentPublicInstance | null>(null);
const wf = inject(injectionKeys.core);
const instancePath = inject(injectionKeys.instancePath);

const { formValue, handleInput } = useFormValueBroker(
	wf,
	instancePath,
	rootInstance,
);
</script>

<style scoped>
@import "@/renderer/sharedStyles.css";
@import "@/renderer/colorTransformations.css";

.CoreRadioInput {
	width: fit-content;
	max-width: 100%;
}

.options {
	display: flex;
	flex-direction: column;
	gap: 5px;
}

.options.horizontal {
	flex-direction: row;
	flex-wrap: wrap;
}

.option {
	display: flex;
	align-items: center;
	color: var(--primaryTextColor);
	font-size: 0.75rem;
}

label,
input {
	cursor: pointer;
}

input {
	margin-right: 8px;
	accent-color: var(--accentColor);
	outline-color: var(--softenedAccentColor);
}
</style>
