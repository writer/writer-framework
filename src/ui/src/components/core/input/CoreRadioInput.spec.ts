import { describe, expect, it, vi } from "vitest";

import CoreRadioInput from "./CoreRadioInput.vue";
import { flushPromises, mount } from "@vue/test-utils";
import injectionKeys from "@/injectionKeys";
import { ref } from "vue";
import { buildMockComponent, buildMockCore, mockProvides } from "@/tests/mocks";

describe("CoreRadioInput", () => {
	it("should render value from the state and forward emit", async () => {
		const { core, userState } = buildMockCore();
		userState.value = { key: "b" };

		core.addComponent(
			buildMockComponent({
				handlers: { "wf-option-change": "python_handler" },
				binding: {
					eventType: "",
					stateRef: "key",
				},
			}),
		);

		const wrapper = mount(CoreRadioInput, {
			global: {
				provide: {
					...mockProvides,
					[injectionKeys.core as symbol]: core,
					[injectionKeys.evaluatedFields as symbol]: {
						label: ref("This is the label"),
						orientation: ref("horizontal"),
						options: ref({ a: "Option A", b: "Option B" }),
					},
				},
			},
		});

		await flushPromises();

		const options = wrapper.findAll("input");
		expect(options).toHaveLength(2);

		const optionA = options.at(0);
		expect(optionA.attributes().value).toBe("a");
		expect(optionA.element.checked).toBe(false);

		const optionB = options.at(1);
		expect(optionB.attributes().value).toBe("b");
		expect(optionB.element.checked).toBe(true);

		const dispatchEvent = vi.spyOn(wrapper.vm.$el, "dispatchEvent");
		await optionA.trigger("input");
		await flushPromises();
		expect(dispatchEvent).toHaveBeenCalledOnce();

		expect(wrapper.element).toMatchSnapshot();
	});
});
