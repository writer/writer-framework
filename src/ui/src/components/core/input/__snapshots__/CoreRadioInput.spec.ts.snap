// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoreRadioInput > should render value from the state and forward emit 1`] = `
<div
  class="BaseInputWrapper CoreRadioInput"
  data-v-7e12caaf=""
  data-v-d075b11d=""
>
  <label
    data-v-7e12caaf=""
  >
    This is the label
  </label>
  
  <div
    class="options horizontal"
    data-v-d075b11d=""
  >
    
    <div
      class="option"
      data-v-d075b11d=""
    >
      <input
        data-v-d075b11d=""
        id="component-id-test:0-option-a"
        name="component-id-test:0-options"
        type="radio"
        value="a"
      />
      <label
        data-v-d075b11d=""
        for="component-id-test:0-option-a"
      >
        Option A
      </label>
    </div>
    <div
      class="option"
      data-v-d075b11d=""
    >
      <input
        data-v-d075b11d=""
        id="component-id-test:0-option-b"
        name="component-id-test:0-options"
        type="radio"
        value="b"
      />
      <label
        data-v-d075b11d=""
        for="component-id-test:0-option-b"
      >
        Option B
      </label>
    </div>
    
  </div>
  
</div>
`;
